### PR TITLE
[IOTDB-1083] There is no need to get write lock each time when we try…

### PR DESCRIPTION
**Pull request by yuqi1129**
… to get or create time series file processor

### Content3 ...

<hr>

This PR has:
- [ ] been self-reviewed.
    - [ ] concurrent read
    - [ ] concurrent write
    - [ ] concurrent read and write 
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. 
- [ ] added or updated version, __license__, or notice information
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious 
  for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold 
  for code coverage.
- [ ] added integration tests.
- [ ] been tested in a test IoTDB cluster.

<hr>

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java [hunk 1]
```diff
@@ -969,35 +969,35 @@ private TsFileProcessor getOrCreateTsFileProcessorIntern(long timeRangeId,
...
-            >= IoTDBDescriptor.getInstance().getConfig().getConcurrentWritingTimePartition()) {
-          Map.Entry<Long, TsFileProcessor> processorEntry = tsFileProcessorTreeMap.firstEntry();
-          logger.info(
-              "will close a {} TsFile because too many active partitions ({} > {}) in the storage group {},",
-              sequence, tsFileProcessorTreeMap.size(),
-              IoTDBDescriptor.getInstance().getConfig().getConcurrentWritingTimePartition(),
-              storageGroupName);
-          asyncCloseOneTsFileProcessor(sequence, processorEntry.getValue());
-        }
+    TsFileProcessor res = tsFileProcessorTreeMap.get(timeRangeId);
```
Comment (jixuan1989) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java:972]: Should consider carefully whether `workSequenceTsFileProcessors` (The `tsFileProcessorTreeMap` in this method) is thread safety.

Comment (yuqi1129) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java:972]: Method `getOrCreateTsFileProcessorIntern` is called by `getOrCreateTsFileProcessor` which ready guard by lock in `StorageGroupProcessor`,  this method in fact is trying to get the same lock again

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java [hunk 2]
```diff
@@ -1006,50 +1006,49 @@ private TsFileProcessor getOrCreateTsFileProcessorIntern(long timeRangeId,
...
     VersionController versionController = getVersionControllerByTimePartitionId(timePartitionId);
+    TsFileProcessor tsFileProcessor = getTsFileProcessor(sequence, filePath, versionController);
+    tsFileProcessor.addCloseFileListeners(customCloseFileListeners);
+    tsFileProcessor.addFlushListeners(customFlushListeners);
+
+    tsFileProcessor.setTimeRangeId(timePartitionId);
+    return tsFileProcessor;
+  }
+
+  private TsFileProcessor getTsFileProcessor(boolean sequence, String filePath,
```
Comment (jixuan1989) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java:1029]: `newTsFileProcess()` is better.
And,  I think the following should also be in the method.
```
    tsFileProcessor.addCloseFileListeners(customCloseFileListeners);
    tsFileProcessor.addFlushListeners(customFlushListeners);

    tsFileProcessor.setTimeRangeId(timePartitionId);
```

Comment (yuqi1129) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java:1029]: Good idea, i will follow your suggestion

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java [hunk 3]
```diff
@@ -969,87 +969,86 @@ private TsFileProcessor getOrCreateTsFileProcessorIntern(long timeRangeId,
...
 
-        // build new processor
-        TsFileProcessor newProcessor = createTsFileProcessor(sequence, timeRangeId);
-        tsFileProcessorTreeMap.put(timeRangeId, newProcessor);
-        tsFileManagement.add(newProcessor.getTsFileResource(), sequence);
-        res = newProcessor;
-      }
+    //Use double-check to shorten the lock range
+    if (null == res) {
+      writeLock();
```
Comment (SilverNarcissus) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java:976]: As you say, outer code has lock this write lock. How come you lock it twice?  May be this lock op can also be removed.

**Review by SilverNarcissus — CHANGES_REQUESTED** [review]
A fantastic observation. But there is a issue and conflict, please have a look.

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java (hunk 3 above)
Comment (yuqi1129) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/StorageGroupProcessor.java:976]: Yes, further optimization about the lock can be made, i will remove related duplication code later, thanks

**Comment by yuqi1129** [comment]
@qiaojialin Hi, i have update the PR, could you take some time to review it ?

**Review by qiaojialin — APPROVED** [review]
LGTM~

**Review by SilverNarcissus — APPROVED** [review]
Look good to me~

**Pull request merged**